### ISSUE TITLE
[Bug] Dark Reader tries to access resources bloked by content blockers, leading to access of undefined or null resources

### ISSUE THREAD

**Issue by bershanskiy** [darkreader/darkreader#6381]
# Bug Report

## Bug Description
Content blocking extensions like uBlock Origin, Ad Block, Ad Block Plus, and others can block resources after they were loaded by network system but before they enter the DOM. Dark Reader does not account for this and tries to access resources which are undefined or null, logging various errors.

There are multiple different manifestations of this bug, but they always come after error `Failed to load resource: net::ERR_BLOCKED_BY_CLIENT` in tab's console.

This bugs results in the following errors in the extension console (incomplete list):
 - `Uncaught TypeError: Cannot read property 'cssRules' of null`
 - `Uncaught TypeError: Cannot read property 'whenDefined' of null` in `addUndefinedResolver()`
 - `Uncaught (in promise) TypeError: Failed to fetch` in `getImageDetails()`

## Steps To Reproduce
I have not found a way to reproduce this bug reliably on the first try. Here is a method which produces an error within ~1 minute:

1. Check out current `master` branch and apply the following rudimentarry patch to log evidence of the race
2. Load extension into the browser and open the extension background console/errors page and look for error
3. Load a large complex site like GitHub or Reddit and keep doing stuff and reloading it untill error is logged
5. Once the error is logged, open site console and look for log messages which were added by the patch.
5. If you could not reproduce the error, delete extension and go to step 2.

```diff
diff --git a/src/inject/dynamic-theme/style-manager.ts b/src/inject/dynamic-theme/style-manager.ts
index 58714093..16585569 100644
--- a/src/inject/dynamic-theme/style-manager.ts
+++ b/src/inject/dynamic-theme/style-manager.ts
@@ -9,6 +9,8 @@ import {bgFetch} from './network';
 import {createStyleSheetModifier} from './stylesheet-modifier';
 import {isShadowDomSupported, isSafari, isThunderbird, isChromium} from '../../utils/platform';

+let counter = 0;
+
 declare global {
     interface HTMLStyleElement {
         sheet: CSSStyleSheet;
@@ -325,14 +327,17 @@ export function manageStyle(element: StyleElement, {update, loadingStart, loadin
         function buildOverrides() {
             const force = forceRenderStyle;
             forceRenderStyle = false;
+            const count = counter;
+            counter++;
             sheetModifier.modifySheet({
-                prepareSheet: prepareOverridesSheet,
+                prepareSheet: () => {console.log('prepareSheet called', count); return prepareOverridesSheet() },
                 sourceCSSRules: rules,
                 theme,
                 ignoreImageAnalysis,
                 force,
                 isAsyncCancelled: () => cancelAsyncOperations,
             });
+            console.log('After sheetModifier.modifySheet', count);
             isOverrideEmpty = syncStyle.sheet.cssRules.length === 0;
             if (sheetModifier.shouldRebuildStyle()) {
                 // "update" function schedules rebuilding the style
```

## Expected Behavior

No errors being logged.

No errors being logged, console messages appear in order:

![image](https://user-images.githubusercontent.com/45960703/127568769-45f8113c-cf13-4352-9b16-b759e2cc7740.png)

## Actual Behavior

Only "After sheetModifier.modifySheet" message is logged, followed by error.
![image](https://user-images.githubusercontent.com/45960703/127569181-9be6f50a-505e-479c-9a8d-3d80d7c46339.png)

![image](https://user-images.githubusercontent.com/45960703/129681165-8c52e953-718a-462c-98ba-dbff82bd9f79.png)

## System Information

- OS: Windows 10 
- Browser: Chrome 92 
- Dark Reader Version: current `master`

## Additional Context

There are no clear visible regressions for the user, aside from, possibly flash of white content before extension changes the site to dark.

**Comment by Gusted** [comment]
Ah nice catch. 

So it's possible that https://github.com/darkreader/darkreader/blob/master/src/inject/dynamic-theme/stylesheet-modifier.ts#L108 is most likely `return` here(which actually shouldn't). As otherwise their is a 100% guarantee the syncStyle is available to be called. Even though the fix for this problem is very simple. I'd like to know which syncStyle is empty.

Because the prepareSheet will be callled. Could you manually debug the content script and see why/what is the culprit(or even adding a check to see if the `return` is the problem. As I wasn't able to reproduce the problem.

**Comment by bershanskiy** [comment]
I found root cause of this bug, will draft a PR fixing it.

**Comment by bershanskiy** [comment]
This bug is not directly related to MV3, but it prevents me from doing regression testing, so I added it to "Manifest V3 Migration" project anyway.

**Comment by bershanskiy** [comment]
I fixed up two places which were causing this bug, and it no longer produces flash of white content. There is a third variant of this bug which dies not appear to have any adverse effects besides logged error. I'm keeping this open untill I fix the last (hopefully) bug variant.

**Comment by Gusted** [comment]
I'm really unsure what this has to do with this specific issue.

**Comment by richardlawley** [comment]
It's another reproducible case leading to a request for undefined (though not related to a content blocker). I did state that it might be a separate bug in that a CSS rule which should be ignored is not.

**Comment by Gusted** [comment]
> I did state that it might be a separate bug in that a CSS rule which should be ignored is not.

That would be indeed another bug, but not something able to fix as we're not a full fledged CSS/DOM parser to know when to load certain things and when not. The request to undefined should be fixed by #7049

**Comment by bershanskiy** [comment]
I am no longer able to reproduce this bug and I believe it has been fixed, closing. Feel free to re-open.
